### ISSUE TITLE
[web] add FragmentProgram support to CanvasKit

### ISSUE THREAD

**Issue by jonahwilliams** [flutter/flutter#114121]
The fragment program API allows users to define GLSL shader variants as assets, load these on demand into their app, and provide them to the Paint.shader property. To date, this has not ever been implemented for the Canvaskit backend.

There are a number of changes that need to be made to both the engine and tooling.

### impellerc

The shader compiler can target either a platform specific shader (for impeller) or SkSL source for a Skia backend. This shader is bundled with metadata that describes the uniform layout in a flatbuffer format. For the web implementation, we'll want to use the SkSL output. For the metadata, we have a few choices:

  * Bundle a generated flat buffer decoder from dart (https://google.github.io/flatbuffers/flatbuffers_guide_use_dart.html ) Ideally this would _not_ need additional dependencies
  * Drop the flatbuffer format for the web and use something like JSON for which there is already a parser.  If we use a different output format, this will need to be specified with a new option flag the tool will pass.

### Flutter tool

Currently the flutter tool disables shader compilation and bundling when targeting web. We'll need to re-enable this, as well as pass any additional flags required for impellerc to ensure they are compiled correctly. Additionally, we may need to do more work to ensure that shader hot restart works properly in both modes.

### Web Engine

The FragmentProgram API requires the usage of a SkRuntimeEffect object. Currently, this interface is only available on the locally-built CanvasKit, and not the bundled canvaskit. We have a few options here:
  * complete the transiton to locally built canvaskit
  * Use a runtime check to see if the RuntimeEffect construct is available
  * switch the upkg version to a canvaskit version with RuntimeEffect
  
Separately, the canvaskit version of the RuntimeEffect API does not seem to accept sampler arguments. We need to follow up with the Skia team to see if this is possible to add. **If not, we may decide not to implement this API to avoid leaving it in a half-working state**

### Web Local Engine Builds

The addition of an impellerc dependency means that the wasm_release engine build will not have all of the artifacts required to run locally. We could temporarily work around this by letting the flutter tool fall back to a local impellerc. Long term, we may move to a more mobile like setup where there is a separate host build that may or may not be specified directly.

**Comment by jonahwilliams** [comment]
The engine side of the hacks required to ship this are collected in https://github.com/flutter/engine/pull/37065

**Comment by jonahwilliams** [comment]
Flatbuffers can generate dart reader/writer. The blobcat schema generates the following file: https://gist.github.com/jonahwilliams/3b4ff0c40d96962304732fd1471bae91

The challenge with including this in the engine is the package dependency.

**Comment by jonahwilliams** [comment]
This flag would be useful if dart supported that:

```
  --gen-onefile          Generate a single output file for C#, Go, Java, Kotlin
                         and Python. Implies --no-include.
```

Could add it, or use JSON instead

**Comment by jonahwilliams** [comment]
https://github.com/flutter/flutter/issues/114260

**Comment by BrianCraig** [comment]
Hi @jonahwilliams,

creating a `FragmentProgram` calling `fromAsset` throws outside a Future, do you consider this change is positive, before this PR closes? (considering also #114260)

https://github.com/flutter/engine/blob/00995b7b910d9d9c412441f02ada13b9aff6f964/lib/web_ui/lib/src/engine/canvaskit/renderer.dart#L388

```dart
if (!isRuntimeEffectAvailable) {
  throw Exception('FragmentProgram is not supported.');
}
```
 to

```dart
if (!isRuntimeEffectAvailable) {
  return Future.error(Exception('FragmentProgram is not supported.'));
}
```

This is in reference to an open issue #117532

Brian
